lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -18,6 +18,9 @@
  *        E.g. match having class `1<"2` needs to recognize `class="1&lt;&quot;2"`.
  * @TODO: Decode character references in `get_attribute()`
  * @TODO: Properly escape attribute value in `set_attribute()`
+ * @TODO: Add slow mode to escape character entities in CSS class names?
+ *        (This requires a custom decoder since `html_entity_decode()`
+ *        doesn't handle attribute character reference decoding rules.
```

Comment:
Do you have an example? A it seems to decode sequences like ‘&#x042’ or ‘&#x2029’

Reply:
Gotcha:
```
php > echo html_entity_decode('&#x042');
&#x042
```

Reply:
I think something is wrong with that snippet as the output for that line should be `B` (and it is for me in a PHP shell).

to borrow an example from the HTML spec:

> If the markup contains (not in an attribute) the string `I'm &notit; I tell you`, the character reference is parsed as "not", as in, `I'm ¬it; I tell you` (and this is a parse error). But if the markup was `I'm &notin; I tell you`, the character reference would be parsed as "notin;", resulting in `I'm ∉ I tell you` (and no parse error).
>
> However, if the markup contains the string `I'm &notit; I tell you` in an attribute, no character reference is parsed and string remains intact (and there is no parse error).

![Screen Shot 2022-09-27 at 12 13 27 PM](https://user-images.githubusercontent.com/5431237/192615423-e013b851-940b-4adc-a735-7d3fb63fd74c.png)

And this is where PHP will mistranslate because it doesn't offer a flag (at least in any function I've found) which indicates the context in which a named character reference is being decoded.

Reply:
> I think something is wrong with that snippet as the output for that line should be B (and it is for me in a PHP shell).

Interesting! I checked on PHP 8.1 and 7.4 and both outputted `&#042`.

Either way your argument makes sense :+1:

Reply:
> Interesting! I checked on PHP 8.1 and 7.4 and both outputted `&#042`.

In review I realize I inserted the missing semicolon in your example code, not having noticed it was missing. After fixing that I can confirm I'm seeing the same behavior your are, and also that the browser adds the missing semicolon (which means it diverges from how PHP handles it, and how the HTML spec suggests it ought to be handled).

<img width="310" alt="Screen Shot 2022-09-28 at 10 32 04 AM" src="https://user-images.githubusercontent.com/5431237/192850486-cefa76b1-a650-4ef3-8f53-ef79c76c0d33.png">

In ambiguous cases for obvious reasons it continues to consume the digits if it can.

<img width="312" alt="Screen Shot 2022-09-28 at 10 38 05 AM" src="https://user-images.githubusercontent.com/5431237/192851194-04c0fa2d-23f3-469e-87be-9173ce690cea.png">

And if the number exceeds the Unicode code point range it does what we expect from the spec and replaces it with a replacement symbol.

<img width="366" alt="Screen Shot 2022-09-28 at 10 38 18 AM" src="https://user-images.githubusercontent.com/5431237/192851262-2480bb1d-3e40-448f-a78f-73680e533a6b.png">